src/avx512bw/avx512bw-from-valid-utf8.inl.cpp

```diff
@@ -0,0 +1,339 @@
+        |aaaa.abbb|bbb?.????|????.???0|0000.0000| 2-byte char -- by 9
+        |aaaa.aaa?|????.????|????.????|?000.0000| ASCII char -- by 7 */
+    {
+        /** pshufb
+
+        continuation = 0
+        ascii    = 7
+        _2_bytes = 9
+        _3_bytes = 10
+        _4_bytes = 11
+
+        shift_left_v3 = 4 * [
+            ascii, # 0000
+            ascii, # 0001
+            ascii, # 0010
+            ascii, # 0011
+            ascii, # 0100
+            ascii, # 0101
+            ascii, # 0110
+            ascii, # 0111
+            continuation, # 1000
+            continuation, # 1001
+            continuation, # 1010
+            continuation, # 1011
+            _2_bytes, # 1100
+            _2_bytes, # 1101
+            _3_bytes, # 1110
+            _4_bytes, # 1111
+        ] */
+        const __m512i shift_left_v3 = _mm512_setr_epi64(
+            0x0707070707070707,
+            0x0b0a090900000000,
+            0x0707070707070707,
+            0x0b0a090900000000,
+            0x0707070707070707,
+            0x0b0a090900000000,
+            0x0707070707070707,
+            0x0b0a090900000000
+        );
+
+        const __m512i shift = _mm512_shuffle_epi8(shift_left_v3, char_class);
+        values = _mm512_sllv_epi32(values, shift);
+    }
+
+    /* 5. Shift right the values by variable amounts to reset lowest bits
+        |0000.0000|000a.aabb|bbbb.cccc|ccdd.dddd| 4-byte char -- by 11
+        |0000.0000|0000.0000|aaaa.bbbb|bbcc.cccc| 3-byte char -- by 16
+        |0000.0000|0000.0000|0000.0aaa|aabb.bbbb| 2-byte char -- by 21
+        |0000.0000|0000.0000|0000.0000|0aaa.aaaa| ASCII char -- by 25 */
+    {
+        // 4 * [25, 25, 25, 25, 25, 25, 25, 25, 0, 0, 0, 0, 21, 21, 16, 11]
+        const __m512i shift_right = _mm512_setr_epi64(
+            0x1919191919191919,
+            0x0b10151500000000,
+            0x1919191919191919,
+            0x0b10151500000000,
+            0x1919191919191919,
+            0x0b10151500000000,
+            0x1919191919191919,
+            0x0b10151500000000
+        );
+
+        const __m512i shift = _mm512_shuffle_epi8(shift_right, char_class);
+        values = _mm512_srlv_epi32(values, shift);
+    }
+
+    return values;
+}
+
+
+// See: http://0x80.pl/notesen/2021-12-22-test-and-clear-bit.html
+bool test_and_clear_bit(uint32_t& val, int bitpos) {
+    const uint32_t bitmask = uint32_t(1) << bitpos;
+    const uint32_t old = val;
+
+    val &= ~bitmask;
+
+    return val < old;
+}
+
+/*
+    utf32_to_utf16 converts `count` lower UTF-32 words
+    from input `utf32` into UTF-16.
+
+    Returns how many 16-bit words were stored.
+*/
+size_t utf32_to_utf16(__m512i utf32, unsigned int count, char16_t* output) {
+    const __mmask16 valid = uint16_t((1 << count) - 1);
+    // 1. check if we have any surrogate pairs
+    const __mmask16 sp_mask = _mm512_mask_cmpgt_epu32_mask(valid, utf32, v_0000_ffff);
+    if (sp_mask == 0) {
+        // XXX: Masked vmovdqa is slow;
```

Comment:
Intel tells us that it has 5 cycles of latency. It is going to be hard to beat by much if that's accurate?

Reply:
Fun fact: Intel reports 5 cycles of latency too for _mm256_storeu_epi16 (unmasked!).

Reply:
So it is conceivable that the masked store is super efficient.